File: backend/internal/trip/trip.go
package trip

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
)

var (
	ErrTripNotFound     = errors.New("trip not found")
	ErrTripNotStarted   = errors.New("trip has not started")
	ErrTripNotCompleted = errors.New("trip is not completed")
	ErrTripCompleted    = errors.New("trip already completed")
	ErrTripCancelled    = errors.New("trip already cancelled")
)

type Status string

const (
	StatusAssigned   Status = "assigned"
	StatusInProgress Status = "in_progress"
	StatusCompleted  Status = "completed"
	StatusCancelled  Status = "cancelled"
)

type SettlementStatus string

const (
	SettlementUnsettled     SettlementStatus = "unsettled"
	SettlementCashCollected SettlementStatus = "cash_collected"
)

type Settlement struct {
	Status          SettlementStatus
	Method          *string
	CashCollectedAt *time.Time
}

type OperationContext struct {
	DriverName   string    `json:"driver_name"`
	VehicleID    uuid.UUID `json:"vehicle_id"`
	Make         string    `json:"make"`
	Model        string    `json:"model"`
	ModelYear    int       `json:"model_year"`
	Color        string    `json:"color"`
	LicensePlate string    `json:"license_plate"`
	ServiceCode  string    `json:"service_code"`
	ServiceName  string    `json:"service_name"`
	Fare         Fare      `json:"fare"`
}

type Fare struct {
	AmountMinor int64  `json:"amount_minor"`
	Currency    string `json:"currency"`
}

type Trip struct {
	OperationContext *OperationContext
	RideRequestID    uuid.UUID
	RiderUserID      uuid.UUID
	DriverUserID     uuid.UUID
	Status           Status
	AssignedAt       time.Time
	StartedAt        *time.Time
	CompletedAt      *time.Time
	CancelledAt      *time.Time
	Settlement       Settlement
}

type Repository interface {
	Start(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error)
	Complete(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error)
	ConfirmCashCollected(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error)
}

type Service struct{ repository Repository }

func NewService(repository Repository) Service { return Service{repository: repository} }

func (s Service) Start(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error) {
	return s.repository.Start(ctx, rideRequestID, driverUserID)
}

func (s Service) Complete(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error) {
	return s.repository.Complete(ctx, rideRequestID, driverUserID)
}

func (s Service) ConfirmCashCollected(ctx context.Context, rideRequestID, driverUserID uuid.UUID) (Trip, error) {
	return s.repository.ConfirmCashCollected(ctx, rideRequestID, driverUserID)
}

File: mobile/lib/features/ride_flow/domain/trip.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'marketplace_request.dart';

part 'trip.freezed.dart';
part 'trip.g.dart';

@freezed
abstract class SettlementSnapshot with _$SettlementSnapshot {
  const factory SettlementSnapshot({
    required String status,
    String? method,
    @JsonKey(name: 'cash_collected_at') DateTime? cashCollectedAt,
  }) = _SettlementSnapshot;

  factory SettlementSnapshot.fromJson(Map<String, dynamic> json) =>
      _$SettlementSnapshotFromJson(json);
}

@freezed
abstract class TripOperationFare with _$TripOperationFare {
  const factory TripOperationFare({
    @JsonKey(name: 'amount_minor') required int amountMinor,
    required String currency,
  }) = _TripOperationFare;

  factory TripOperationFare.fromJson(Map<String, dynamic> json) =>
      _$TripOperationFareFromJson(json);
}

@freezed
abstract class TripOperationContext with _$TripOperationContext {
  const factory TripOperationContext({
    @JsonKey(name: 'vehicle_id') required String vehicleId,
    @JsonKey(name: 'service_code') required String serviceCode,
    @JsonKey(name: 'service_name') required String serviceName,
    @JsonKey(name: 'driver_name') required String driverName,
    required String make,
    required String model,
    @JsonKey(name: 'model_year') required int modelYear,
    required String color,
    @JsonKey(name: 'license_plate') required String licensePlate,
    required TripOperationFare fare,
  }) = _TripOperationContext;

  factory TripOperationContext.fromJson(Map<String, dynamic> json) =>
      _$TripOperationContextFromJson(json);
}

@freezed
abstract class TripSnapshot with _$TripSnapshot {
  const factory TripSnapshot({
    @JsonKey(name: 'operation_context') TripOperationContext? operationContext,
    @JsonKey(name: 'ride_request_id') String? rideRequestId,
    @JsonKey(name: 'rider_user_id') String? riderUserId,
    @JsonKey(name: 'driver_user_id') String? driverUserId,
    RideLocation? pickup,
    RideLocation? destination,
    required String status,
    @JsonKey(name: 'assigned_at') required DateTime assignedAt,
    @JsonKey(name: 'started_at') DateTime? startedAt,
    @JsonKey(name: 'completed_at') DateTime? completedAt,
    @JsonKey(name: 'cancelled_at') DateTime? cancelledAt,
    required SettlementSnapshot settlement,
  }) = _TripSnapshot;

  factory TripSnapshot.fromJson(Map<String, dynamic> json) =>
      _$TripSnapshotFromJson(json);
}

@freezed
abstract class DriverTripHistoryItem with _$DriverTripHistoryItem {
  const factory DriverTripHistoryItem({
    @JsonKey(name: 'operation_context') TripOperationContext? operationContext,
    @JsonKey(name: 'ride_request_id') required String rideRequestId,
    required RideLocation pickup,
    required RideLocation destination,
    required String status,
    @JsonKey(name: 'assigned_at') required DateTime assignedAt,
    @JsonKey(name: 'started_at') DateTime? startedAt,
    @JsonKey(name: 'completed_at') DateTime? completedAt,
    @JsonKey(name: 'cancelled_at') DateTime? cancelledAt,
    required SettlementSnapshot settlement,
  }) = _DriverTripHistoryItem;

  factory DriverTripHistoryItem.fromJson(Map<String, dynamic> json) =>
      _$DriverTripHistoryItemFromJson(json);
}
